src/extensions/banks_client.rs

```diff
@@ -230,51 +243,189 @@ impl BanksClientExtensions for BanksClient {
+        .await?;
+
+        return Ok(associated_token_account);
+    }
+
+    async fn deploy_program(
+        &mut self,
+        path_to_program: &str,
+        program_keypair: &Keypair,
+        payer: &Keypair,
+    ) -> transport::Result<()> {
+        let (buffer, buffer_len) = util::load_file_to_bytes(path_to_program);
+
+        let program_data = buffer;
+
+        // multiply by 2 so program can be updated later on
+        let program_len = buffer_len;
+        let minimum_balance = Rent::default().minimum_balance(
+            bpf_loader_upgradeable::UpgradeableLoaderState::programdata_len(program_len)
+                .expect("Cannot get program len"),
         );
+        let latest_blockhash = self.get_latest_blockhash().await?;
 
-        self.process_transaction(
-            Transaction::new_signed_with_payer(
-                &[ix],
-                Some(&payer.pubkey()),
-                &[payer],
-                latest_blockhash
+        // 1 Create account
+        self.process_transaction(system_transaction::create_account(
+            &payer,
+            &program_keypair,
+            latest_blockhash,
+            minimum_balance,
+            program_len as u64,
+            &bpf_loader::id(),
+        ))
+        .await
+        .unwrap();
+
+        // 2. Write to buffer
+        let deploy_ix = |offset: u32, bytes: Vec<u8>| {
+            loader_instruction::write(&program_keypair.pubkey(), &bpf_loader::id(), offset, bytes)
+        };
+
+        let chunk_size =
+            util::calculate_chunk_size(&deploy_ix, &vec![payer, program_keypair], latest_blockhash);
+
+        for (chunk, i) in program_data.chunks(chunk_size).zip(0..) {
+            let ix = deploy_ix(i * chunk_size as u32, chunk.to_vec());
+            let message = Message::new(&[ix], Some(&payer.pubkey()));
+            let tx = Transaction::new(&[payer, program_keypair], message, latest_blockhash);
+            self.process_transaction(tx).await?;
+        }
+
+        // 3. Finalize
+        let finalize_msg = Message::new_with_blockhash(
+            &[loader_instruction::finalize(
+                &program_keypair.pubkey(),
+                &bpf_loader::id(),
+            )],
+            Some(&payer.pubkey()),
+            &latest_blockhash,
+        );
+        let finalize_tx =
+            Transaction::new(&[payer, program_keypair], finalize_msg, latest_blockhash);
+        self.process_transaction(finalize_tx).await?;
+
+        return Ok(());
+    }
+
+    async fn deploy_upgradable_program(
+        &mut self,
+        path_to_program: &str,
+        buffer_keypair: &Keypair,
+        buffer_authority_signer: &Keypair,
+        program_keypair: &Keypair,
+        payer: &Keypair,
+    ) -> transport::Result<()> {
+        let (buffer, buffer_len) = util::load_file_to_bytes(path_to_program);
+
+        let program_data = buffer;
+
+        // multiply by 2 so program can be updated later on
+        let program_len = buffer_len * 2;
```

Comment:
I know it's not a big deal in test environment but his doubles the rent. Since on upgrade the entire binary has to be uploaded I think we can keep the actual `buffer_len` and on redeploy close the account, reallocate the space required and then uploaded the program

Reply:
@lowprivuser yes and no, this is how the official Solana CLI `deploy` command behaves... https://github.com/solana-labs/solana/blob/master/cli/src/program.rs#L987 

So the question is if we'd like to optimize it on our own or follow what Solana CLI is doing... WDYT?

Reply:
interesting, let's stick to what CLI does for now then